lib/receiver/receiver_impl.cc

```diff
@@ -196,10 +196,15 @@ namespace gr
 	size_t inputs_to_process = d_cell_allocation.size();
 	if (d_process_uplink)
         	inputs_to_process *= 2;
-	for(int input_nr=0;input_nr<inputs_to_process;input_nr++){
-		boost::thread t1(&receiver_impl::synchronized_handler,this,input, input_items, noutput_items, input_nr);
-		usleep(200);
+	boost::thread *threads[inputs_to_process];	      
+	for(int input_nr=1;input_nr<inputs_to_process;input_nr++){
+		class receiver_impl *r1=  new receiver_impl(d_OSR, d_cell_allocation, d_tseq_nums, d_process_uplink);
+		boost::thread t1(&receiver_impl::synchronized_handler,r1,input, input_items, noutput_items, input_nr);
```

Comment:
I know I've recommended not passing `this` to the threads but this is not going to work either because each instance will have its own members but the `synchronized_handler` needs the values detected from the C0 channel.
Try passing the `this` instead but protect the members with a mutex.
The alternative is before starting the threads to set r1 members to the values from the `this` object.

Reply:
but...any idea why BCCH data from C0 stops appearing?

Reply:
No :(
Which of the attached files contains C0 and what is the sampling rate 0.2M?

Reply:
arfcn 77

Reply:
Yes sample rate of 200KSPS.

Reply:
Could you share the flowgraph you are testing with?
With grgsm_decode compiled from the PR I'm able get BCCH decoded from arfcn77
```
$ grgsm_decode -c arfcn77 -a 77 -s 0.2M -v
1429664 2208770:  15 06 21 00 01 f0 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b
1429668 2208902:  2d 06 3f 00 4a e0 4d 1b b2 0c 01 00 0b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b
1429674 2209100:  25 06 21 00 05 f4 11 30 77 04 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b
1429678 2209232:  2d 06 3f 00 4a e0 4d 1b b2 0c 01 00 0b 2b 2b 2b 2b 2b 2b 2b 2b 2b 2b
```
But it does not work if I don't specify `-a 77` so it needs frequency correction

Reply:
But do you get data out in hopping condition.

Reply:
I'm not testing the hopping mode as I don't expect the multi-threaded receiver to fix the problem with hopping channels, if it works correctly it's only going to increase the performance and hopefully you will be able to process all 16 channels in realtime.
Also the files you've attached contain only encrypted calls so I can't test hopping without having the Kc keys.

Reply:
No the calls aren't encrypted, they are A5/0 calls. Also by " I don't expect the multi-threaded receiver to fix the problem with hopping channels", can you please specify issues?? I thought only handling data in real-time was the problem.

Reply:
> No the calls aren't encrypted, they are A5/0 calls

I was looking only on SDCCH8 in ts1 and I found the 2 unencrypted Assignment Commands on ts2.
It's going to be a lot easier if you share the grc file you use for testing together with the exact commands as I'm not going to spent hours constructing my own.

> Also by " I don't expect the multi-threaded receiver to fix the problem with hopping channels", can you please specify issues?? 

Maybe I'm mixing threads as there were some reports about problems with hopping channels. Anyway if your problem is not related with the hopping mode please rename the PR to avoid confusion.

Reply:
Also don't forget that you need to implement one of the two options described in https://github.com/ptrkrysik/gr-gsm/pull/412#discussion_r194834443 without them there is no any point in testing other channels (hopping or not)

Reply:
Hopping grc
[grc_hopping_grc.zip](https://github.com/ptrkrysik/gr-gsm/files/2098353/grc_hopping_grc.zip)

Reply:
You are not storing the threads in the `threads` array so the join result in a segfault.
```
Thread 100 "receiver2" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff2a7c4700 (LWP 22478)]
0x00007fffed8cc166 in boost::thread::get_thread_info() const () from /lib64/libboost_thread.so.1.63.0
(gdb) bt
#0  0x00007fffed8cc166 in boost::thread::get_thread_info() const () from /lib64/libboost_thread.so.1.63.0
#1  0x00007fffed8cd111 in boost::thread::native_handle() () from /lib64/libboost_thread.so.1.63.0
#2  0x00007fffde020c36 in boost::thread::get_id (this=0x7fffde02a42a <std::allocator<gr::tag_t>::allocator()+24>)
    at /usr/include/boost/thread/detail/thread.hpp:742
#3  0x00007fffde020c98 in boost::thread::join (this=0x7fffde02a42a <std::allocator<gr::tag_t>::allocator()+24>)
    at /usr/include/boost/thread/detail/thread.hpp:763
#4  0x00007fffde06d2fa in gr::gsm::receiver_impl::work (this=0x5555562f92f0, noutput_items=2067, 
    input_items=std::vector of length 16, capacity 16 = {...}, output_items=std::vector of length 0, capacity 0)
    at /home/vasko/sources/gr-gsm/gr-gsm/lib/receiver/receiver_impl.cc:219
```

Reply:
The value returned on line  238 is wrong
```cpp
 238       return d_samples_consumed;
```
it's value is set here
```cpp
 530         if (input_nr == input_items.size() - 1) {
 531           /* Go to the next burst */
 532           d_burst_nr++;
 533 
 534           /* Consume samples of the burst up to next guard period */
 535           to_consume += TS_BITS * d_OSR + d_burst_nr.get_offset();
 536 //          consume_each(to_consume);
 537           d_samples_consumed += to_consume;
 538         }
```
but only when processing the last channel. After I removed lines 530 and 538 in order to increment in each and every channel I started getting BCCH on C0 using your flowgraph.

Reply:
so I need to comment these two..ok doing the commits here..tomorrow i will test these and close the pull-request

Reply:
Vasil

Did you change anything else...because I am not able to get data for C0 in without hopping mode also.

Reply:
Hello Vasil

Was able to get data on C0 in both hopping and non hopping condition. But don't you think the rate of data printing in without hopping mode is more than in hopping mode?

Reply:
How do you test both modes? Give me the exact commands or provide the exact flowgraphs.

Reply:
Grc for single channel with and without hopping.
[grc_without_hopping_and_multiple_hopping.zip](https://github.com/ptrkrysik/gr-gsm/files/2114018/grc_without_hopping_and_multiple_hopping.zip)

Reply:
Vasil

Any update?

Reply:
> But don't you think the rate of data printing in without hopping mode is more than in hopping mode?

Yes it's slower, far from real time and on my laptop it's even slower then with the single threaded receiver (although I could test it only on my laptop that does not have 16/32 cores to do a real test). The CPU cores utilization is still not good which indicates you have a lot more work to do
```
top - 22:20:20 up 4 days, 22:04,  1 user,  load average: 0,87, 0,97, 1,07
Tasks: 295 total,   1 running, 230 sleeping,   0 stopped,   0 zombie
%Cpu0  : 95,7 us,  2,1 sy,  0,0 ni,  0,0 id,  0,0 wa,  2,1 hi,  0,0 si,  0,0 st
%Cpu1  :  6,5 us,  2,2 sy,  0,0 ni, 91,3 id,  0,0 wa,  0,0 hi,  0,0 si,  0,0 st
%Cpu2  : 10,9 us,  2,2 sy,  0,0 ni, 87,0 id,  0,0 wa,  0,0 hi,  0,0 si,  0,0 st
%Cpu3  : 13,0 us,  2,2 sy,  0,0 ni, 84,8 id,  0,0 wa,  0,0 hi,  0,0 si,  0,0 st
KiB Mem : 16117172 total,  1292156 free,  8503096 used,  6321920 buff/cache
KiB Swap:  3870716 total,  3864956 free,     5760 used.  8514268 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2425 vasko     20   0 8707900 190820  41908 S 119,1  1,2   0:20.51 python2

...................

top - 22:20:32 up 4 days, 22:04,  1 user,  load average: 0,89, 0,97, 1,07
Tasks: 295 total,   3 running, 229 sleeping,   0 stopped,   0 zombie
%Cpu0  : 27,7 us,  1,0 sy,  0,0 ni, 70,3 id,  0,0 wa,  0,7 hi,  0,3 si,  0,0 st
%Cpu1  : 67,4 us,  1,3 sy,  0,0 ni, 30,9 id,  0,0 wa,  0,3 hi,  0,0 si,  0,0 st
%Cpu2  :  8,6 us,  2,0 sy,  0,0 ni, 88,7 id,  0,0 wa,  0,3 hi,  0,3 si,  0,0 st
%Cpu3  : 22,2 us,  2,3 sy,  0,0 ni, 74,8 id,  0,0 wa,  0,3 hi,  0,3 si,  0,0 st
KiB Mem : 16117172 total,  1262380 free,  8533044 used,  6321748 buff/cache
KiB Swap:  3870716 total,  3864956 free,     5760 used.  8484552 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2425 vasko     20   0 10,008g 219072  41968 S 113,0  1,4   0:34.44 python2
```
My advice is to profile the application in order to find where the most of the time is spent.

In your current implementation new N - 1 threads are started on every invocation of `receiver_impl::work` function and these threads run for a very short time as they consume (`d_samples_consumed`) only 625 samples from the input (`noutput_items`). Try to rewrite the code so the receiver thread together with the N - 1 sub-threads consume as much of the input as possible as starting and stopping threads is an expensive operation. Or better try to start them only once and then pass them the input through a queue.

Reply:
or can we do something like this:
Currently I am using gnuradio c++ api to create flow graphs instead of gnuradio-companion. So With every value of chaning MAI, I will launch a new grc which will read only one arfcn at a time?

Reply:
I'm not sure whether I correctly understand your idea but I'm highly skeptical about it. My advice is still to profile and improve the multi-threaded support.

Reply:
Suppose I trigger all the threads once in the initialization of the constructor, then what about synchronization, only for C0 synchronization has to be done and not for the reaming hopping channels. Also you mentioned to consumed samples more than 625, its an argument of the work function, how do I control that?

Reply:
Print the values of the argument (noutput_items) and the return value (d_samples_consumed) and you will see what I mean. And yes, you will still need to synchronize the threads somehow.

Reply:
OK..If I understand you clearly you want me to add a buffer, say a circular buffer which can hold some amount of samples inside work function and all the threads read from the buffer. Is it correct?

Reply:
When threads are long living you need to find a way to pass the samples between the main receiver thread and the sub threads and each thread should receive only the samples for its channel and not all samples. It's up to you to decided how to implement this.

Reply:
Understood...thanks!!!

Reply:
hello Vasil

Was able to handle hopping in realtime, but am not getting anything out of control channel decoder in wireshark, but if i  connect wireshark just after timeslot demapper, I can see the LAPdm messages, any idea where I am going wrong?

Reply:
> any idea where I am going wrong?

No.